### PR TITLE
port/MycroftAI/lingua-franca/pull/232

### PR DESCRIPTION
**Pull request by NeonJarbas**
port https://github.com/MycroftAI/lingua-franca/pull/232

(note, basque was already localized in #3 )

## Summary by CodeRabbit

- **New Features**
  - Introduced `nice_relative_time` function for generating human-readable time differences in the app.

- **Documentation**
  - Updated notes in the `nice_relative_time_eu` function to specify time reporting limitations.

- **Tests**
  - Added test cases for `nice_relative_time` to validate various time differentials.

### CODE REVIEW

**Review by NeonDaniel — COMMENTED** [review]
I'm not sure I agree with always rounding to one unit; if I request something in "2 hours and 15 minutes" replying with "in 2 hours" would indicate to me that the "and 15 minutes" was mis-transcribed..

Maybe this could optionally accept a "precision" param? Caller could specify 'hours'/'minutes'/'seconds', etc.

**Comment by ChanceNCounter** [comment]
> Maybe this could optionally accept a "precision" param? Caller could specify 'hours'/'minutes'/'seconds', etc.

Unported work from mycroft.util.lang introduced the relevant enum, but I don't think anyone ever used it

File: lingua_franca/format.py [hunk 1]
```diff
@@ -578,3 +580,56 @@ def nice_response(text, lang=''):
...
+    delta = to_local(when) - now
+
+    if delta.total_seconds() < 1:
+        return "now"
+
+    if delta.total_seconds() < 90:
+        if delta.total_seconds() == 1:
+            return "one second"
+        else:
+            return "{} seconds".format(int(delta.total_seconds()))
```
[the remark this answers is not in the document]
Comment (krisgesling) [reply lingua_franca/format.py:614]: Yeah, I think the current implementation is the simplest format that might be useful, but it still depends on the specific use case. Definitely lots of room for improvement here.

I'm also interested to look at how this relates to `nice_duration()` which takes a number of seconds as input but is in some ways another end of the same spectrum - it allows you to get a very specific representation of a period of time. Maybe a precision param provides that nice middle ground?

**Comment by krisgesling** [comment]
> I'm not sure I agree with always rounding to one unit; if I request something in "2 hours and 15 minutes" replying with "in 2 hours" would indicate to me that the "and 15 minutes" was mis-transcribed..
> 
> Maybe this could optionally accept a "precision" param? Caller could specify 'hours'/'minutes'/'seconds', etc.

Yeah, I think the current implementation is the simplest format that might be useful, but it still depends on the specific use case. Definitely lots of room for improvement here.

I'm also interested to look at how this relates to `nice_duration()` which takes a number of seconds as input but is in some ways another end of the same spectrum - it allows you to get a very specific representation of a period of time. Maybe a precision param provides that nice middle ground?

**Comment by NeonDaniel** [comment]
> Unported work from mycroft.util.lang introduced the relevant enum, but I don't think anyone ever used it

Didn't think to look, but I do see that now..
https://github.com/MycroftAI/mycroft-core/blob/1b67bf33e08541c34c331e7980ce0d6b4bc9c708/mycroft/util/format.py#L53

> I'm also interested to look at how this relates to `nice_duration()` which takes a number of seconds as input but is in some ways another end of the same spectrum - it allows you to get a very specific representation of a period of time. Maybe a precision param provides that nice middle ground?

`nice_relative_time`, `nice_duration`

Maybe it would make sense to deal with total seconds and some default format in both cases? Something like:

```python
def _get_spoken_duration(total_seconds: int, precision: TimeResolution):
    # Do formatting here

def nice_relative_time(when: datetime, relative_to: datetime):
    if relative_to:
        now = relative_to
    else:
        now = now_local()
    delta = to_local(when) - now
    if delta.total_seconds() < 1:
        return "now"

    if delta.total_seconds() < 90:
        if delta.total_seconds() == 1:
            return "one second"
        else:
            return _get_spoken_duration(delta.total_seconds(), TimeResolution.SECONDS
    ...
    
def nice_duration(duration, lang='', speech=True, precision: TimeResolution=TimeResolution.SECONDS):
    days = int(duration // 86400)
    hours = int(duration // 3600 % 24)
    minutes = int(duration // 60 % 60)
    seconds = int(duration % 60)

    if speech:
        out = ""
        if days > 0 and precision>=TimeResolution.DAYS:
    ...
```

**Comment by JarbasAl** [comment]
@coderabbitai review

**Comment by JarbasAl** [comment]
@coderabbitai review

**Pull request closed without merging**